File: Java/22-GenerateParentheses.java
public class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new LinkedList<>();
        generate(0, 0, n, result, "");
        return result;
    }
    
    public void generate(int open, int close, int n, List<String> res, String curr){
        if (open == n && close == n){
            //all of the parenthesis are distributed - adding a valid distribution
            res.add(curr);
        }else if (open == n){
            //all of the opening parenthesis are distributed - no other 
            //option than to add the remaining closing parenthesis
            while (close < n){
                curr += ")";
                close ++;
            }
            res.add(curr);
        }else if (close > open){
            //invalid distribution - backtracking
            return;
        }else{
            //on this step it is possible to either add a closing or an opening parenthesis
            generate(open + 1, close, n, res, curr + "(");
            generate(open, close + 1, n, res, curr + ")");
        }
        
    }
}

File: cpp/22-GenerateParentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        if (n == 0) {
            res.push_back("");
            return res;
        }
        generate(n, res, "", 0, 0);
        return res;
    }

    void generate(int n, vector<string>& res, string curr, int open, int close) {
        if (open == n && close == n) {
            res.push_back(curr);
        } else if (open == n && close < n) {
            while (close < n) {
                curr += ")";
                close++;
            }
            res.push_back(curr);
        } else if (close > open) {
//      not right below!!
//        } else if (open != n && close == n) {
            return;
        } else {
            generate(n, res, curr + "(", open + 1, close);
            generate(n, res, curr + ")", open, close + 1);
        }
    }
};
